sql/parser/builtins.go

```diff
@@ -62,103 +80,94 @@ type builtin struct {
 // Need to figure out if there is a standard set of functions or if we can
 // customize this to our liking. Would be good to support type conversion
 // functions.
-var builtins = map[string]builtin{
+var builtins = map[string][]builtin{
 	// TODO(XisiHuang): support encoding, i.e., length(str, encoding).
-	"length": stringBuiltin(func(s string) (Datum, error) {
+	"length": {stringBuiltin(func(s string) (Datum, error) {
 		return DInt(len(s)), nil
-	}),
+	})},
 
-	"lower": stringBuiltin(func(s string) (Datum, error) {
+	"lower": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToLower(s)), nil
-	}),
+	})},
 
-	"upper": stringBuiltin(func(s string) (Datum, error) {
+	"upper": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
-	}),
+	})},
 
-	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax.
-	"substring": {
-		nArgs: -1,
+	"substr":    substringImpls,
+	"substring": substringImpls,
+
+	// concat concatenates the text representations of all the arguments.
+	// NULL arguments are ignored.
+	"concat": {
+		builtin{
+			fn: func(args DTuple) (Datum, error) {
+				var buffer bytes.Buffer
+				for _, d := range args {
+					ds, err := datumToRawString(d)
+					if err != nil {
+						return DNull, err
+					}
+					buffer.WriteString(ds)
+				}
+				return DString(buffer.String()), nil
+			},
+		},
+	},
+}
+
+var substringImpls = []builtin{
+	{
+		types: typeList{stringType, intType},
 		fn: func(args DTuple) (Datum, error) {
-			argsNum := len(args)
-			if argsNum != 2 && argsNum != 3 {
-				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
-					argsNum, "2 or 3")
+			str := string(args[0].(DString))
```

Comment:
Did my comment about not casting to `string` get lost? You should be able to slice a `DString`.